### PR TITLE
[Lora] Support long context lora 

### PR DESCRIPTION
**Pull request by rkooo567**
Currently we need to call rotary embedding kernel for each LoRA, which makes it hard to serve multiple long context length LoRA. Add batched rotary embedding kernel and pipe it through.

It replaces the rotary embedding layer to the one that is aware of multiple cos-sin-cache per scaling factors. 

Follow up of https://github.com/vllm-project/vllm/pull/3095/files

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/lora/layers.py [hunk 1]
```diff
@@ -1193,3 +1203,204 @@ def can_replace_layer(cls, source_layer: nn.Module,
                           model_config: Optional[PretrainedConfig]) -> bool:
         # Special handling for the LogitsProcessor.
         return False
+
+
+class LoRALinearScalingRotaryEmbedding(LinearScalingRotaryEmbedding):
```
Comment (rkooo567) [inline vllm/lora/layers.py:1208]: not needed

File: vllm/lora/layers.py [hunk 2]
```diff
@@ -1193,3 +1203,204 @@ def can_replace_layer(cls, source_layer: nn.Module,
...
+        offsets: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Use specialized kernel that indexes int the cos/sin cache.
+        ops.batched_rotary_embedding(positions, query, key, self.head_size,
+                                     self.cos_sin_cache, self.is_neox_style,
+                                     self.rotary_dim, offsets)
+        return query, key
+
+
+class LoRAPagedAttentionWithRoPE(LoRALayer):
```
Comment (rkooo567) [inline vllm/lora/layers.py:1276]: not needed

File: tests/lora/conftest.py [hunk 3]
```diff
@@ -21,6 +21,41 @@
 from vllm.model_executor.layers.vocab_parallel_embedding import ParallelLMHead
 from vllm.model_executor.model_loader import get_model
 
+LONG_LORA_INFOS = [
```
Comment (rkooo567) [inline tests/lora/conftest.py:24]: Will be moved to hf hub

**Comment by rkooo567** [comment]
The PR is ready to be reviewed

**Comment by richardliaw** [comment]
tests failing?

**Comment by rkooo567** [comment]
that' expected because we need to move fine tuned lora model to hf hub so that CI can access it. 

will be done soon. The test succeeds locally.

**Review by rkooo567 — COMMENTED** [review]
@Yard1 compared to internal version;
- clean up on some of variables and logic
- Added a test to test_layers.py
- Replaced rotary embedding instead of self.attn
- fixed a bug where rotary emb is only replacing the first layer
- download lora fine tuned models from hf hub I created

File: vllm/core/scheduler.py [hunk 4]
```diff
@@ -650,11 +650,14 @@ def _schedule_prefills(
                 num_prompt_tokens = waiting_seqs[0].get_len()
                 assert num_new_tokens == num_prompt_tokens
 
-            if num_new_tokens > self.prompt_limit:
+            prompt_limit = (seq_group.lora_request.long_lora_max_len
+                            if seq_group.lora_request
+                            and seq_group.lora_request.long_lora_max_len else
+                            self.prompt_limit)
```
Comment (Yard1) [inline vllm/core/scheduler.py:656]: let's put it in a method/function

File: vllm/engine/arg_utils.py [hunk 5]
```diff
@@ -397,6 +398,10 @@ def add_cli_args(
             choices=['auto', 'float16', 'bfloat16', 'float32'],
             help=('Data type for LoRA. If auto, will default to '
                   'base model dtype.'))
+        parser.add_argument('--long-lora-scaling-factors',
+                            type=nullable_str,
+                            default=EngineArgs.long_lora_scaling_factors,
+                            help='Scaling factors of long LoRAs')
```
Comment (Yard1) [inline vllm/engine/arg_utils.py:404]: "Specify multiple scaling factors (which can be different from base model scaling factor - see eg. Long LoRA) to allow for multiple LoRA adapters trained with those scaling factors to be used at the same time. If not specified, only adapters trained with the base model scaling factor are allowed."

File: vllm/engine/output_processor/stop_checker.py [hunk 6]
```diff
@@ -59,7 +65,9 @@ def maybe_stop_sequence(self, seq: Sequence, new_char_count: int,
             return
 
         # Check if the sequence has reached max_model_len.
-        if seq.get_len() > self.max_model_len:
+        max_len = (lora_req.long_lora_max_len if lora_req
+                   and lora_req.long_lora_max_len else self.max_model_len)
```
Comment (Yard1) [inline vllm/engine/output_processor/stop_checker.py:69]: make a method

File: vllm/lora/layers.py [hunk 7]
```diff
@@ -1193,3 +1200,101 @@ def can_replace_layer(cls, source_layer: nn.Module,
                           model_config: Optional[PretrainedConfig]) -> bool:
         # Special handling for the LogitsProcessor.
         return False
+
+
+class LinearScalingRotaryEmbeddingWithLora(BaseLayerWithLoRA):
+    """Implements RoPE-scaled embeddings with linear scaling for
+    multiple LoRA adapters with a specialized kernel.
+
+    Replace LinearScalingRotaryEmbedding to MultiLinearScalingRotaryEmbedding
```
Comment (Yard1) [inline vllm/lora/layers.py:1209]:
```suggestion
    Replace LinearScalingRotaryEmbedding with MultiLinearScalingRotaryEmbedding
```

File: vllm/transformers_utils/tokenizer_group/tokenizer_group.py [hunk 8]
```diff
@@ -39,15 +39,33 @@ def encode(self,
...
                lora_request: Optional[LoRARequest] = None) -> List[int]:
         tokenizer = self.get_lora_tokenizer(lora_request)
-        return tokenizer.encode(prompt)
+        ret = tokenizer.encode(prompt)
+        input_length = len(ret)
+        if lora_request:
+            max_input_length = (lora_request.long_lora_max_len
+                                or self.max_input_length)
+        else:
+            max_input_length = self.max_input_length
```
Comment (Yard1) [inline vllm/transformers_utils/tokenizer_group/tokenizer_group.py:48]: make a method/funcion

File: tests/lora/test_lora_checkpoints.py [hunk 9]
```diff
@@ -28,6 +28,7 @@ def test_load_checkpoints(
         # and the test should pass.
         LoRAModel.from_local_checkpoint(
             baichuan_lora_files,
+            4096,
```
Comment (Yard1) [inline tests/lora/test_lora_checkpoints.py:31]: use kwarg instead

File: tests/lora/test_lora_manager.py [hunk 10]
```diff
@@ -293,7 +293,7 @@ def test_lru_cache_worker_lora_manager(llama_2_7b_model_extra_embeddings,
                                        sql_lora_files):
     lora_config = LoRAConfig(max_lora_rank=8, max_cpu_loras=4, max_loras=4)
     worker_lora_manager = LRUCacheWorkerLoRAManager(
-        4, 2, llama_2_7b_model_extra_embeddings.unpadded_vocab_size -
+        4, 2, 8192, llama_2_7b_model_extra_embeddings.unpadded_vocab_size -
```
Comment (Yard1) [inline tests/lora/test_lora_manager.py:296]: use kwargs instead

**Review by Yard1 — COMMENTED** [review]
Thanks, looks good! Some small comments

File: tests/lora/test_lora_checkpoints.py (hunk 9 above)
Comment (rkooo567) [reply tests/lora/test_lora_checkpoints.py]: Default None == meaning it will use scaling factor 1. Also added docstring

File: tests/lora/test_lora_manager.py (hunk 10 above)
Comment (rkooo567) [reply tests/lora/test_lora_manager.py]: handled same ways as above

**Comment by rkooo567** [comment]
@Yard1 everything addressed

**Review by Yard1 — APPROVED** [review]

**Comment by AllenDou** [comment]
May I ask a question? Only LinearScalingRotaryEmbedding requires the 'withlora' version. Do DynamicNTKScalingRotaryEmbedding/YaRNScalingRotaryEmbedding/Phi3SuScaledRotaryEmbedding also require it?

**Comment by rkooo567** [comment]
I think those are not working with long context multi lora. In order to get it working, I think other rotary embedding should also support multi scaling factors like we are doing for `LinearScalingRotaryEmbedding `.

**Comment by AllenDou** [comment]
ok, thanks for your response :)
